### PR TITLE
ShaderTrap Examples: reduction example by bitwise and, or and xor

### PR DESCRIPTION
**Pull request by sliu-UIUC**
Addresses part of #70 by adding reduction examples using bitwise and, or and xor operations.

### CODE REVIEW

File: examples/OpenGL45/arrayReductionByBitwiseAnd_version2.shadertrap [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+BIND_SHADER_STORAGE_BUFFER BUFFER partial_bitwise_and BINDING 1
+BIND_SHADER_STORAGE_BUFFER BUFFER result BINDING 2
+
+DECLARE_SHADER shader KIND COMPUTE
+#version 450
+
+layout(local_size_x=1, local_size_y=1, local_size_z=1) in;
+
+layout(std430, binding = 0) buffer input_array {
+  uint[256] data1;
```
Comment (afd) [inline examples/OpenGL45/arrayReductionByBitwiseAnd_version2.shadertrap]: Call this data

File: examples/OpenGL45/arrayReductionByBitwiseAnd_version2.shadertrap [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+};
+
+layout(std430, binding = 1) buffer partial_bitwise_and {
+  uint[16] data2;
```
Comment (afd) [inline examples/OpenGL45/arrayReductionByBitwiseAnd_version2.shadertrap]: Call this partial_results

File: examples/OpenGL45/arrayReductionByBitwiseAnd_version3.shadertrap [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  uint id = uint(gl_LocalInvocationID.x);
+  // depth of the tree where we perform reduction. 
+  uint Depth = findMSB(data.length());
+  
+  // At each step, half of invocations drop out of the computation. 
+  // Index corresponding to active invocations will update with bitwise and operation while
+  // the rest will set to corresponding id in case race condition between multiple threads occurs.
+  // This way each position in the input array other than index 0 should be set to its index value
+  // by the corresponding invocation.  
+  // Each active invocation will add its value to that of a neighboring invocatiion.
```
Comment (afd) [inline examples/OpenGL45/arrayReductionByBitwiseAnd_version3.shadertrap]: "add" isn't right here, becuase this is doing bitwise and, not addition.

File: examples/OpenGL45/arrayReductionByBitwiseAnd_version5.shadertrap [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+shared uint shared_memory[256];
+
+void main() {
+ 
+  // The partition size is the number of elements in data to be handled for each work group. 
+  uint partitionSize = data.length() / gl_NumWorkGroups[0];
+  
+  // Depth of reduction tree for the current partition.  
+  uint Depth = findMSB(partitionSize);
```
Comment (afd) [inline examples/OpenGL45/arrayReductionByBitwiseAnd_version5.shadertrap]: I suggest using snake_case for variable names in GLSL, so partition_size, and depth.

File: examples/OpenGL45/arrayReductionByBitwiseOr_version1.shadertrap [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+};
+
+layout(std430, binding = 1) buffer result {
+  uint res;
+};
+
+void main() {
+	for(int i = 0; i < 256; ++i) {
+		res |= data[i];
+	}
```
Comment (afd) [inline examples/OpenGL45/arrayReductionByBitwiseOr_version1.shadertrap]: I suggest using 2-space indentation. At least, be consistent in how much indentation you use throughout the PR.

File: examples/OpenGL45/arrayReductionByBitwiseOr_version2.shadertrap [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+#version 450
+
+layout(local_size_x=1, local_size_y=1, local_size_z=1) in;
+
+layout(std430, binding = 0) buffer input_array {
+  uint[256] data1;
+};
+
+layout(std430, binding = 1) buffer partial_bitwise_or {
+  uint[16] data2;
```
Comment (afd) [inline examples/OpenGL45/arrayReductionByBitwiseOr_version2.shadertrap]: Same comments as above about the names of the data1 and data2 fields.

File: examples/OpenGL45/arrayReductionByBitwiseOr_version2.shadertrap [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+layout(std430, binding = 2) buffer result {
+  uint res;
+};
+
+void main() {
+	for(uint i = gl_WorkGroupID[0]*16; i < gl_WorkGroupID[0]*16 + 16; ++i) {
+		data2[gl_WorkGroupID[0]] |= data1[i];
+	}
+	atomicOr(res, data2[gl_WorkGroupID[0]]);
```
Comment (afd) [inline examples/OpenGL45/arrayReductionByBitwiseOr_version2.shadertrap]: Again, please be consistent with indentation - worth checking carefully throughout the PR.

File: examples/OpenGL45/arrayReductionByBitwiseOr_version3.shadertrap [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+  uint id = uint(gl_LocalInvocationID.x);
+  // depth of the tree where we perform reduction. 
+  uint Depth = findMSB(data.length());
+  
+  // At each step, half of invocations drop out of the computation. 
+  // Index corresponding to active invocations will update with bitwise or operation while
+  // the rest will set to corresponding id in case race condition between multiple threads occurs.
+  // This way each position in the input array other than index 0 should be set to its index value
+  // by the corresponding invocation.  
+  // Each active invocation will add its value to that of a neighboring invocatiion.
```
Comment (afd) [inline examples/OpenGL45/arrayReductionByBitwiseOr_version3.shadertrap]: "Add" is not right.

File: examples/OpenGL45/arrayReductionByBitwiseOr_version5.shadertrap [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+  uint id = uint(gl_LocalInvocationID.x) + n_prev;
+  
+  // Data is copied over to shared memory by 256 threads in a work group.
+  shared_memory[id] = data[id];
+  barrier();
+  
+  for(uint i = 1; i <= Depth; ++i) {
+  	if(id % (1 << i) == 0) shared_memory[id] |= shared_memory[id + (1<<(i-1))];
+  	barrier();  	 
+  }
```
Comment (afd) [inline examples/OpenGL45/arrayReductionByBitwiseOr_version5.shadertrap]: Looks like a mix of 2- and 4-space indentation.

File: examples/OpenGL45/arrayReductionByBitwiseXor_version1.shadertrap [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+layout(std430, binding = 1) buffer result {
+  uint res;
+};
+
+void main() {
+	for(int i = 0; i < 256; ++i) {
+		res ^= data[i];
+	}
+}
```
Comment (afd) [inline examples/OpenGL45/arrayReductionByBitwiseXor_version1.shadertrap]: Indentation

File: examples/OpenGL45/arrayReductionByBitwiseXor_version2.shadertrap [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+
+layout(local_size_x=1, local_size_y=1, local_size_z=1) in;
+
+layout(std430, binding = 0) buffer input_array {
+  uint[256] data1;
+};
+
+layout(std430, binding = 1) buffer partial_bitwise_xor {
+  uint[16] data2;
+};
```
Comment (afd) [inline examples/OpenGL45/arrayReductionByBitwiseXor_version2.shadertrap:41]: Ditto.

File: examples/OpenGL45/arrayReductionByBitwiseXor_version3.shadertrap [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+  uint id = uint(gl_LocalInvocationID.x);
+  // depth of the tree where we perform reduction. 
+  uint Depth = findMSB(data.length());
+  
+  // At each step, half of invocations drop out of the computation. 
+  // Index corresponding to active invocations will update with bitwise xor operation while
+  // the rest will set to corresponding id in case race condition between multiple threads occurs.
+  // This way each position in the input array other than index 0 should be set to its index value
+  // by the corresponding invocation.  
+  // Each active invocation will add its value to that of a neighboring invocatiion.
```
Comment (afd) [inline examples/OpenGL45/arrayReductionByBitwiseXor_version3.shadertrap]: Ditto.

**Review by afd — CHANGES_REQUESTED** [review]
This looks excellent! My requested changes are all just nits about whitespace, comments or naming.

File: examples/OpenGL45/arrayReductionByBitwiseOr_version1.shadertrap (hunk 5 above)
Comment (sliu-UIUC) [reply examples/OpenGL45/arrayReductionByBitwiseOr_version1.shadertrap]: Indentation issues fixed.

File: examples/OpenGL45/arrayReductionByBitwiseAnd_version2.shadertrap (hunk 2 above)
Comment (sliu-UIUC) [reply examples/OpenGL45/arrayReductionByBitwiseAnd_version2.shadertrap]: done

File: examples/OpenGL45/arrayReductionByBitwiseAnd_version2.shadertrap (hunk 1 above)
Comment (sliu-UIUC) [reply examples/OpenGL45/arrayReductionByBitwiseAnd_version2.shadertrap]: done

File: examples/OpenGL45/arrayReductionByBitwiseAnd_version3.shadertrap (hunk 3 above)
Comment (sliu-UIUC) [reply examples/OpenGL45/arrayReductionByBitwiseAnd_version3.shadertrap]: done

File: examples/OpenGL45/arrayReductionByBitwiseAnd_version5.shadertrap (hunk 4 above)
Comment (sliu-UIUC) [reply examples/OpenGL45/arrayReductionByBitwiseAnd_version5.shadertrap]: Done

File: examples/OpenGL45/arrayReductionByBitwiseOr_version3.shadertrap (hunk 8 above)
Comment (sliu-UIUC) [reply examples/OpenGL45/arrayReductionByBitwiseOr_version3.shadertrap]: Done

**Review by afd — APPROVED** [review]
Thanks for this contribution!

**Pull request merged**